### DISCUSSION TITLE
STAC v1.9

### DISCUSSION THREAD

**Discussion by mqshaw** [Xylopyrographer/STAC#35] [General]
@Xylopyrographer 

> Well, at least we made some progress with 1.9!
> 
> In one way, the errors in the serial monitor log are a good thing. Might be the easiest to fix. Looking at line 296, in the getTallyState() function:
> 
> if(stClient.connect(stIP, stPort, 100)) {
> 
> the timeout value of 100ms in that call corresponds to the call returning as seen in the error log because it timed out. So it looks like the link is “connected” but the host hasn’t returned a socket within the 100ms allocated to it. It may be as simple then as finding an appropriate timeout value to use in the stClient.connect() call. To make that easier, the startup code could be modified so you could input that value in the serial monitor during a restart.
> 
> If the Arduino-esp32 folks can fix the “client disconnect on test after buffer read" bug that would go a long way to resolving this issue as well since the connection should then remain persistent.
> 
> Might be useful to turn on the “Show timestamp” checkbox on the serial monitor window. Would let us know the time between events.

I will do that on Sunday.

> On your test system at home, do you see purple X’s at the same rate if running the python 2 version of the Smart Tally simulator?

I have not tried Python 2 version on my simulator. I can try that using the original python script.

> At church, are you live-streaming or doing NDI video over the same WiFi network that the Roland switch is connected to?

Our setup is the following ... 
- The church has a main wired network, with a network switch that we plug into from our router.
- The RiverHouse router supports both 5G and 2.4G Wifi, and wired connections
- The V60-HD is wired into the Router
- The STACs are connected to the RiverHouse router
- The video encoder ( we use a Teradek VidiU Pro ) is hard wired into the RiverHouse router, and we stream directly to YouTube over the main wired network.

> Found the issue causing the display glitches in v1.9.
> 
> Adding the line:
> 	WiFi.setSleep(false);
> back into the connectToWifi() function fixed it up.
> 
> I actually had that line in there since the beginning of time. Not sure what possessed me to remove it…
> 
> Regardless, the init code in the WiFi connect routine is back to what it was in v1.8 and earlier with the added warning comment to myself:
> 
>     WiFi.disconnect(true);
>     WiFi.mode(WIFI_STA);                        // set WiFi station mode
>     WiFi.setSleep(false);                       // have to set this mode or else the display goes nuts when polling
> 
> I’ve released v1.9.1 into the main branch and revised the Release Notes accordingly on GitHub.
> 
> Not having that line in there really messes up the machine. For fun, with v1.9 loaded, and the STAC communicating and polling to the server, press and hold the button until the change brightness mode comes up. Then just watch as the STAC auto iterates through the levels. I think the entire RTOS gets whacked waiting for the WiFi to come out of sleep mode. Or something.
> 
> I’ll let you take v1.9.1 for a spin before making an official v1.9.1 release. Let me know when you think it’s OK to do that.
> 

OK, I will have it up and running on Sunday at church.  I will run both the base 1.9.1 version and the version with timeout tracking added.

**Comment by mqshaw** [comment]
@Xylopyrographer 

I ran the 1.9.1 code on Sunday night. The output of the serial is here.  As you can see, there are still some cases where the STAC would have shown a purple X, but the timeout tracking code prevents it.

[STAC 1.9.txt](https://github.com/Xylopyrographer/STAC/files/6576823/STAC.1.9.txt)

Thanks
Mark

**Reply by Xylopyrographer** [reply]
Interesting isn't it. Do you think it'd be worthwhile playing with the time-out value in the call to connect? Wonder how close we are to the 100ms? Might help to set a new "standard" value?

Cheers,
    Rob
' ===================

**Reply by mqshaw** [reply]
Yeah, it might take trial and error to find out.  The problem is that the longer the time-out, the less responsive the STAC will be.

**Reply by Xylopyrographer** [reply]
I think we'd be OK with making that magic number whatever we'd like as I believe the function timeout value is the _maximum_ it'll wait for a connection. It returns if the connection is made in less than that time.

Be easy enough to test. Set it to some ridiculously high value and confirm it doesn't actually wait that long.

**Comment by Xylopyrographer** [comment]
Good day Mark. Everything look OK to you to do an official 1.9.1 release later today or tomorrow?

**Reply by mqshaw** [reply]
Hi Rob,
Yes, I think that 1.9.1 is ready to go.  
I do think that we should integrate the timeout tracking, it solves the spurious X issue. But I am happy to incorporate that into a new major version.

**Reply by Xylopyrographer** [reply]
OK. I'll release it tomorrow.

Agree with the timeout stuff. We can do a point release for that or roll it into the "next big thing".

**Comment by Xylopyrographer** [comment]
@mqshaw
Hi there Mark. Hope you're enjoying the long weekend. Using VSCode as the IDE, created (eventually) a new branch `version2-development`.

Could have lots of discussion on this, but basically this branch runs on Arduino-core v2.0.3 only due to breaking changes from core v1.0.6. More in the new `Detailed Change Log` doc in the `STAC`  folder.

Also incorporated a somewhat significant rewrite of the error handling stuff at the bottom of `loop()`. This incorporates the "BPX" filtering, but in a better way.

In that, added filtering for all errors as with core v2.0.3 I'm seeing odd stuff coming back from `WiFiClient`. Not sure if this just due to `WiFiClient` or is because of the older python2 STS emulator I'm using for testing or a combination of both or 🤷🏼‍♂️.

Be good if you can to throw this into an ATOM & see how it goes for you. As usual, any feedback would be great.

In learning how VSCode integrates with GitHub, not entirely sure what if anything that did to the branch we checked out during our zoom call.

**Comment by mqshaw** [comment]
Hi Rob,

We should probably merge the current code 1.11-RC1 to master, so we keep trunk clean.

I doubt that you have done anything, and the beauty of Git is that we can undo unintended changes if we find them.

Then we create a new branch off the new master (after 1.11-RC merge), and we can test from there.

How does that sound? I can submit a pull request to get the ball rolling, but you have master locked, so are the only one that can commit the merge.

Mark


Sent from my Mobile Device
________________________________

**Comment by Xylopyrographer** [comment]
Morning Mark:

Agree it's time to do some house cleaning.

At the moment we have eight branches open:
```
main
mgs_STAC 1.11B9
mqs_timeout_fixes 1.10 base
mqs_timeout_tracking_1.9_base
mqs_timeout_tracking_1.9.1_base
mqs_timeout_tracking_21-04-25
V 1.10
version2-development
```
The `1.11-RC1` branch is the one we set up on my local machine on our zoom call. (I believe.)

I made a number of changes to 1.11-RC1 and spawned those into the `version2-development` branch as, in my way of thinking, there isn't going to be a 1.11 release as the latest arduino-esp32 core v2.0.3 has breaking changes incompatible with core v1.0.6. But it does offer a number of improvements and is the way things are going so now seems to be as good a time as any to migrate.

The STAC v2.0 code makes it easier (over the 1.11-RC1 stream) to drop in different filtering schemes for various error conditions.

So, I think it would make sense to PR the `version2-development` branch into `main` and then after it is merged, "delete" what would be the remaining 6 branches?

**Comment by mqshaw** [comment]
Hi Rob,

Agreed, time for some cleanup.

Here is what I would suggest …


1.       We merge [mqs_STAC 1.11B9] back to master (main). On my system this compiled happily using the v1.0.6 core and this is what I have been running for the last 3 weeks.



2.       We then create a label for that version on Master for posterity.

        git --tag -a “STAC V1.11” -m “Final STAC code using Arduino-esp core v1.0.6”



3.       We the delete the following branches, they are no longer needed.



git push origin --delete  mqs_STAC 1.11B9

git push origin --delete  mqs_timeout_fixes 1.10 base

git push origin --delete  mqs_timeout_tracking_1.9_base

git push origin --delete  mqs_timeout_tracking_1.9.1_base

git push origin --delete  mqs_timeout_tracking_21-04-25

The above is a remote delete, if the branches are checked out on your local system you will need to do a local delete

  git branch -d <branch>

Now, it is time to bring in the v2.0.3 changes.

Because version2-development is a branch off of the 1.11-RC1 branch, the merge should be straightforward …  as long as it is done the right way.


  1.  Because Master has changed …  First we must merge master back into the version2-development branch.



git checkout main

  [ Make sure it’s clean and up-to-date ]

git pull

  [ force update to the main repo with any changes, should be none ]

git checkout version2-development

  [ switch to the branch ]

git checkout -b version2-dev_main_merge

  [ create a clean copy of the branch to perform the merge ]



git -merge main –-no-commit

  [ merging changes from main into version2-development branch, without commit ]

  [ resolve any merge conflicts and check before committing changes to branch ]

  https://explainshell.com/explain?cmd=git+merge+-s+ours+--no-ff+--no-commit+-e+dev



  1.  Finally, we can now create a PR to merge the new branch [version2-dev_main_merge]to main.  Which will be clean and easy because we have done all of the work to resolve any conflicts on the [version2-dev_main_merge] branch.  When using GitHub to make and manage the PR, one can choose to delete the merge branch once the merge is complete.


  1.  You can have the [ version2-development] alive until you are comfortable that it can be deleted.

How does this sound?  We can do a zoom to work through this together if you like.

Mark

**Reply by Xylopyrographer** [reply]
zoom call sounds great. Wednesday morning is booked but I'm open otherwise. Did you want to shoot me a link?
